File: 0200-number-of-islands/0200-number-of-islands.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int n = grid.size();
        int m = grid[0].size();
        int count = 0;
        queue<pair<int,int>> q;
        int drow[] = {-1,0,1,0};
        int dcol[] = {0,1,0,-1};
        for(int i =0;i<n;i++){
            for(int j = 0 ; j < m ; j++){
                if(grid[i][j]=='1'){
                    count++;
                    q.push({i,j});
                    grid[i][j] = '0';
                }

                while(!q.empty()){
                    int row = q.front().first;
                    int col = q.front().second;
                    q.pop();

                    for(int k = 0 ; k < 4 ; k++){
                        int nrow = row + drow[k];
                        int ncol = col + dcol[k];
                        if(nrow<n && ncol<m && nrow>=0 && ncol>=0 && grid[nrow][ncol]!='0'){
                            q.push({nrow,ncol});
                            grid[nrow][ncol] = '0';
                        }
                    }
                }
            }
        }
        return count;
    }
};

File: 0200-number-of-islands/0200-number-of-islands.py
from collections import deque
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        n = len(grid)
        m = len(grid[0])

        q = deque()

        drow = [-1,0,1,0]
        dcol = [0,1,0,-1]

        count = 0

        for i in range (n):
            for j in range (m):
                if grid[i][j] == '1':
                    q.append((i,j))
                    grid[i][j] == '0'
                    count = count + 1

                    while q:
                        row = q[0][0]
                        col = q[0][1]
                        q.popleft()

                        for k in range (4):
                            nrow = row + drow[k]
                            ncol = col + dcol[k]
                            if(nrow >= 0 and ncol >= 0 and nrow < n and ncol < m and grid[nrow][ncol] != '0'):
                                q.append((nrow,ncol))
                                grid[nrow][ncol] = '0'
        
        return count
